types/metadataV14.go

```diff
@@ -0,0 +1,458 @@
+package types
+
+import (
+	"fmt"
+	"hash"
+	"strings"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v3/scale"
+	"github.com/centrifuge/go-substrate-rpc-client/v3/xxhash"
+)
+
+// Based on:
+// https://github.com/polkadot-js/api/blob/master/packages/types/src/interfaces/metadata/v14.ts
+type MetadataV14 struct {
+	Lookup    PortableRegistry
+	Pallets   []PalletMetadataV14
+	Extrinsic ExtrinsicV14
+
+	LookUpData map[int64]*Si1Type
+}
+
+type ExtrinsicV14 struct {
+	Type             Si1LookupTypeID
+	Version          U8
+	SignedExtensions []SignedExtensionMetadataV14
+}
+
+type SignedExtensionMetadataV14 struct {
+	Identifier       Text
+	Type             Si1LookupTypeID
+	AdditionalSigned Si1LookupTypeID
+}
+
+func (m MetadataV14) Encode(encoder scale.Encoder) error {
+	err := encoder.Encode(m.Lookup)
+	if err != nil {
+		return err
+	}
+
+	err = encoder.Encode(m.Pallets)
+	if err != nil {
+		return err
+	}
+
+	return encoder.Encode(m.Extrinsic)
+}
+
+func (m *MetadataV14) Decode(decoder scale.Decoder) error {
+	var err error
+	err = decoder.Decode(&m.Lookup)
+	if err != nil {
+		return err
+	}
+
+	m.LookUpData = make(map[int64]*Si1Type)
+	for _, lookUp := range m.Lookup {
+		m.LookUpData[lookUp.ID.Int64()] = &lookUp.Type
```

Comment:
Hello @NunoAlexandre , I found this code, it is a good idea. This can quickly get the corresponding `Si1Type` through map in the `FindEventNamesForEventID` and `FindCallIndex` methods, instead of looping over and over again. But this piece of code is placed in the decode process. During the decode process, `lookUp.Type` is a temporary variable, fetching the pointer and storing it will cause the subsequent value to be overwritten on this temporary variable, and finally `LookUpData ` will have the same final value inside.

Reply:
Good catch! Working on it 👍

Reply:
Fixed in https://github.com/centrifuge/go-substrate-rpc-client/pull/206/commits/4865d8f4b81e29d4e0c3fcf39636108b391ba1ee